### PR TITLE
[SPARK-36711][PYTHON][FOLLOW-UP] Refactor typing logic for multi-index support

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

This PR proposes to refactor typing logic for multi-index support that was mostly introduced in https://github.com/apache/spark/pull/34176

At a high level, the below functions were introduced

```bash
_extract_types # renamed from `extract_types`
```

```
_is_named_params
_address_named_type_hoders
_to_tuple_of_params
_convert_tuples_to_zip
_address_unnamed_type_holders
```

In this PR, they become as below with simplification:

```bash
_to_type_holders # renamed from `_extract_types`
```

```bash
_new_type_holders # merged from `_is_named_params`, etc.
```

### Why are the changes needed?

To make the codes easier to read.

### Does this PR introduce _any_ user-facing change?

No, dev-only.

### How was this patch tested?

Existing tests should cover them.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @ueshin @itholic @dchvn FYI

**Review by dchvn — APPROVED** [review]
Look so much better! LGTM

File: python/pyspark/pandas/typedef/typehints.py [hunk 1]
```diff
@@ -726,131 +729,108 @@ def create_tuple_for_frame_type(params: Any) -> object:
...
+        #   DataFrame[[index_type, ...], [type, ...]]
+        #   DataFrame[dtypes instance, dtypes instance]
+        #   DataFrame[(index_name, index_type), [(name, type), ...]]
+        #   DataFrame[(index_name, index_type), zip(names, types)]
+        #   DataFrame[[(index_name, index_type), ...], [(name, type), ...]]
+        #   DataFrame[zip(index_names, index_types), zip(names, types)]
+        def is_list_of_pairs(p: Any) -> bool:
+            return (
+                isinstance(p, list) and len(p) >= 1 and isinstance(p[0], tuple) and len(p[0]) == 2
+            )
```
[the remark this answers is not in the document]
Comment (HyukjinKwon) [reply python/pyspark/pandas/typedef/typehints.py:757]: Thanks, yeah let's do that.

**Review by itholic — APPROVED** [review]
LGTM if test passes

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**